Split SourceABC to multiple interfaces

m-novikov opened the pull request:
This pull request splits SourceABС into multiple interfaces. Old SourceABC was used in Image, Slice and Data sources.
But they all had different function and different interfaces. To better communicate intent now there is
`DataSourceABC` for querying "raw" data
`Slice2DSourceABC` for slicing data sources and producing 2d arrays.
`ImageSourceABC` for converting 2d arrays into QImage

k-dominik commented on tests/test_utility/test_abc.py:
```diff
@@ -0,0 +1,45 @@
+        with pytest.raises(TypeError):
+            self.ISignal()
+
+    def test_abstract_signal_raises_type_error(self):
```
is this some pytest dark magic that makes multiple methods with the same name work? In any case, I'd prefer not to have that.

k-dominik commented on volumina/pixelpipeline/datasources/interface.py:
```diff
@@ -0,0 +1,61 @@
+    pass
+
+
+class DataRequestABC(ABC):
```
Is there a good reason why this does not reside in `pixelpipeline.asyncabcs`?

m-novikov replied:
Some requests (e.g. GrayscaleImageRequest) don't implement getResult method.

k-dominik replied:
sure, I mean the whole interface - i'd say it qualifies as async?

m-novikov replied:
I am not even sure that `asyncabc` is a reasonable name :) It should be also `interfaces.py` or something

k-dominik commented on tests/test_pixelpipeline/test_slicesources.py:
```diff
@@ -0,0 +1,36 @@
+        self.assertTrue(np.all(sl == self.raw[1, :, :, 127, 2]))
+
+        sl_bounded = self.ss.request((slice(0, 3), slice(1, None))).wait()
+        self.assertTrue(np.all(sl_bounded == self.raw[1, 0:3, 1:, 127, 2]))
```
```suggestion
        numpy.testing.assert_array_equal(sl_bounded, self.raw[1, 0:3, 1:, 127, 2])
```

k-dominik commented on volumina/pixelpipeline/slicesources.py:
```diff
@@ -42,7 +43,7 @@
 
 
-class SliceRequest(object):
+class SliceRequest(RequestABC):
```
```suggestion
class PlanarSliceRequest(RequestABC):
```

k-dominik commented on volumina/pixelpipeline/imagepump.py:
```diff
@@ -351,15 +351,15 @@ def __init__(self, layerStackModel, sliceProjection, sync_along=(0, 1, 2)):
 
     # mappings
     def layerToSliceSources(self, layer):
```
don't know if this should be renamed to be uber consistent

m-novikov replied:
renamed `layerToPlanarSliceSources`

m-novikov commented:
Moved interfaces of pixelpipeline to interface
Removed getResult method.
* results are already cached for every wait method
* calling getResult on not awaited request will return None, which is error-prone as we have temporal coupling here.

k-dominik commented:
> Removed getResult method.

great
